be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -98,116 +120,150 @@ class SinkBuffer {
     }
 
     void add_request(const TransmitChunkInfo& request) {
-        if (_is_finished) {
-            request.params->release_finst_id();
-            return;
+        DCHECK(_expected_eos > 0);
+        {
+            // Guarantee that no request will be added to buffer after finishing stage
+            std::lock_guard<std::mutex> l(_mutex);
+            if (_is_finishing) {
+                request.params->release_finst_id();
+                return;
+            } else {
+                _buffers[request.fragment_instance_id].push(request);
+            }
         }
+
+        _try_to_trigger_io_task();
+    }
+
+    bool is_full() const {
+        // TODO(hcf) if one channel is congested, it may cause all other channel unwritable
+        // std::queue' read is concurrent safe without mutex
+        for (auto& [_, buffer] : _buffers) {
+            if (buffer.size() > config::pipeline_io_buffer_size) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool is_finished() {
+        if (!_is_finishing) {
+            return false;
+        }
+
+        // Here is the guarantee that
+        // 1. No new io task will be created after finishing stage
+        // 2. No new brpc process will be triggered after finishing stage
+        // So we just wait for existed io task and bprc process to be finished
+        return !_is_io_task_active && _in_flight_rpc_num.load(std::memory_order_acquire) == 0;
+    }
+
+private:
+    void _try_to_trigger_io_task() {
+        // Guarantee that no new io task will be created after finishing stage
         std::lock_guard<std::mutex> l(_mutex);
-        _buffers[request.fragment_instance_id].push(request);
-        _buffer_empty_cv.notify_one();
+        if (_is_finishing) {
+            return;
+        }
+
+        // Tell io task scan buffer one more time
+        if (_is_io_task_active) {
+            _need_one_more_scan = true;
```

Comment:
suggest to use _need_one_more_task or _need_one_more_send

Reply:
ok, changed to _need_one_more_check